Add CUDA Sanitizer

sypneiwski opened the pull request:
Example of a simple synchronization error:
```
a = torch.rand(4, 2, device="cuda")

with torch.cuda.stream(second_stream):
    torch.mul(a, 5, out=a)
```
Output produced by CSAN:
```
============================
CSAN detected a possible data race on tensor with data pointer 139719969079296
Access by stream 94646435460352 during kernel:
aten::mul.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
writing to argument: self, out, output
With stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 364, in _handle_kernel_launch
    stack_trace = traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 544, in __torch_dispatch__
    errors = self.event_handler._handle_kernel_launch(
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 9, in <module>
    torch.mul(a, 5, out=a)

Previous access by stream 0 during kernel:
aten::rand(int[] size, *, int? dtype=None, int? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
writing to argument: output
With stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 364, in _handle_kernel_launch
    stack_trace = traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 544, in __torch_dispatch__
    errors = self.event_handler._handle_kernel_launch(
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 6, in <module>
    a = torch.rand(10000, device="cuda")

Tensor was allocated with stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 420, in _handle_memory_allocation
    traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/utils/_cuda_trace.py", line 23, in fire_callbacks
    cb(*args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/_ops.py", line 60, in __call__
    return self._op(*args, **kwargs or {})
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 541, in __torch_dispatch__
    outputs = func(*args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 6, in <module>
    a = torch.rand(10000, device="cuda")
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84428
* #84427
* __->__ #83984

sypneiwski committed:
Add CUDA Sanitizer

[ghstack-poisoned]

sypneiwski committed:
Update on "Add CUDA Sanitizer"




[ghstack-poisoned]

sypneiwski committed:
Update on "Add CUDA Sanitizer"




[ghstack-poisoned]

sypneiwski committed:
Update on "Add CUDA Sanitizer"




[ghstack-poisoned]

sypneiwski committed:
Update on "Add CUDA Sanitizer"




[ghstack-poisoned]

ngimel commented on test/test_cuda_sanitizer.py:
```diff
@@ -0,0 +1,433 @@
+        read_only: List[DataPtr] = None,
+        read_write: List[DataPtr] = None,
+    ) -> None:
+        self.assertNotEqual(self.kernel_launch(stream, read_only, read_write), [])
```
what result do you expect here? I guess it would depend on what you are testing, but you can make expected result an argument to assert to do a more exact check

sypneiwski replied:
I've added an argument to test the number of errors reported.

ngimel commented on test/test_cuda_sanitizer.py:
```diff
@@ -0,0 +1,433 @@
+
+    def test_all_reads_checked_passing(self):
+        iterations = 10
+        for i in range(1, iterations):
```
took me a few moments to notice difference between this and previous test, could you add a comment why the previous one is failing

sypneiwski replied:
Done.

ezyang commented on torch/_cuda_sanitizer/enable.py:
```diff
@@ -0,0 +1,3 @@
+from torch._cuda_sanitizer._sanitizer import CUDASanitizer
+
+cuda_sanitizer = CUDASanitizer()
```
It's better to not use a side-effectful module; otherwise tools that remove dead Python imports will remove this tool.

ezyang commented on torch/_cuda_sanitizer/_sanitizer.py:
```diff
@@ -0,0 +1,490 @@
+import enum
```
nit: if you under-scored the parent module, you don't have to underscore this module. But I'd probably suggest putting this in `torch.cuda._sanitizer`, which makes use of the existing namespace torch.cuda

sypneiwski replied:
Done.

ezyang commented on torch/_cuda_sanitizer/_sanitizer.py:
```diff
@@ -0,0 +1,490 @@
+    _PYTHON_DISPATCH_TRACE_OFFSET,
+)
+from torch.utils._pytree import tree_map
+
```
Would be nice to have a top level comment describing the overall strategy for this file

ezyang commented on torch/_cuda_sanitizer/_sanitizer.py:
```diff
@@ -0,0 +1,490 @@
+    stream: StreamId
+    operator: str
+    names: List[str]
+    stack_trace: List[str]
```
@zdevito you had advice for capturing Python stack traces efficiently, right?

sypneiwski replied:
@zdevito  Thank you for the suggestion!

I also wanted to ask if you knew of a way to retrieve the line from the source code, given only the bytecode + lasti pair. I know that the line number of the trace can be easily retrieved, but wasn't able to find a way of capturing the source code.

@ezyang I thought it might improve readability to also print the source line in the error message, but perhaps the efficiency is more important?

ezyang replied:
I think the code should retain the info; the point is to not spend a lot of time uselessly formatting stack traces when they never get seen by the user.

sypneiwski replied:
I've ended up using the `traceback.StackSummary.extract` method. I looked at its source code, and it performs only the necessary operations for it to be formatted later.

ezyang commented:
I recommend beefing up the commit message a bit. One suggestion is showing what the output looks like for a "real world" example.

ezyang commented on torch/_cuda_sanitizer/_sanitizer.py:
```diff
@@ -0,0 +1,490 @@
+
+
+def format_log_message(message: str) -> str:
+    return " ".join(line.strip() for line in message.strip().splitlines())
```
Hmm, what's going on here?

sypneiwski replied:
This is supposed to de-wrap the log messages so they don't contain indentation and newlines.

ezyang commented on torch/_cuda_sanitizer/_sanitizer.py:
```diff
@@ -0,0 +1,490 @@
+        return "reading from" if self is AccessType.READ else "writing to"
+
+
+@dataclass
```
Comment saying what this class is!  In particular, it would be helpful to know what the semantics of "operator" and "names" are; they are non-obvious from the types.

ezyang commented on torch/_cuda_sanitizer/_sanitizer.py:
```diff
@@ -0,0 +1,490 @@
+
+@dataclass
+class TensorInfo:
+    allocation_trace: Optional[List[str]]
```
`allocation_stack_trace`?

sypneiwski replied:
Done.

ezyang commented on torch/_cuda_sanitizer/_sanitizer.py:
```diff
@@ -0,0 +1,490 @@
+class TensorInfo:
+    allocation_trace: Optional[List[str]]
+    reads: List[Access] = field(default_factory=list)
+    write: Optional[Access] = None
```
A comment for this class would help explicate why there can be multiple reads but only one write

ezyang commented on torch/_cuda_sanitizer/_sanitizer.py:
```diff
@@ -0,0 +1,490 @@
+                    f"""
+                    Found tensor with pointer: {data_ptr}, but no matching tensor
+                    allocation in the trace. Backfilling the trace now.
+                    No action is necessary.
```
So... why do we need this log message then?

sypneiwski replied:
We've decided to log such occurrences, as something like this (missing the allocation event) should never occur during normal execution and could be a result of some obscure error.

ezyang replied:
Then that's what the log message should say :)

sypneiwski replied:
Now that CSAN is enabled by a function, I've added a warning that perhaps it is being enabled too late in the program, as that is now the most probable reason.

ezyang commented on torch/_cuda_sanitizer/_sanitizer.py:
```diff
@@ -0,0 +1,490 @@
+        if errors:
+            for error in errors:
+                print(error)
+            sys.exit()
```
Hmm, you sure you want to do it this way, as opposed to raising an exception?

sypneiwski replied:
I changed it so the errors are printed to stderr and an aggregate exception is thrown.

ezyang commented:
Needs user facing docs

ezyang commented:
Very cool!

sypneiwski committed:
Update on "Add CUDA Sanitizer"




[ghstack-poisoned]

sypneiwski committed:
Update on "Add CUDA Sanitizer"

Example of a simple synchronization error:
```
a = torch.rand(4, 2, device="cuda")

with torch.cuda.stream(second_stream):
    torch.mul(a, 5, out=a)
```
Output produced by CSAN:
```
============================
CSAN detected a possible data race on tensor with data pointer 139719969079296
Access by stream 94646435460352 during kernel:
aten::mul.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
writing to argument: self, out, output
With stack trace:
  File "/private/home/sypniewski/pytorch/tester.py", line 10, in <module>
    torch.mul(a, 5, out=a)

Previous access by stream 0 during kernel:
aten::rand(int[] size, *, int? dtype=None, int? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
writing to argument: output
With stack trace:
  File "/private/home/sypniewski/pytorch/tester.py", line 7, in <module>
    a = torch.rand(4, 2, device="cuda")

Tensor was allocated with stack trace:
  File "/private/home/sypniewski/pytorch/tester.py", line 7, in <module>
    a = torch.rand(4, 2, device="cuda")
```




[ghstack-poisoned]

sypneiwski commented:
The last commit adds requested comments.

sypneiwski committed:
Update on "Add CUDA Sanitizer"

Example of a simple synchronization error:
```
a = torch.rand(4, 2, device="cuda")

with torch.cuda.stream(second_stream):
    torch.mul(a, 5, out=a)
```
Output produced by CSAN:
```
============================
CSAN detected a possible data race on tensor with data pointer 139719969079296
Access by stream 94646435460352 during kernel:
aten::mul.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
writing to argument: self, out, output
With stack trace:
  File "/private/home/sypniewski/pytorch/tester.py", line 10, in <module>
    torch.mul(a, 5, out=a)
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 486, in __torch_dispatch__
    errors = self.event_handler._handle_kernel_launch(
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 331, in _handle_kernel_launch
    stack_trace = traceback.format_stack()

Previous access by stream 0 during kernel:
aten::rand(int[] size, *, int? dtype=None, int? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
writing to argument: output
With stack trace:
  File "/private/home/sypniewski/pytorch/tester.py", line 7, in <module>
    a = torch.rand(4, 2, device="cuda")
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 486, in __torch_dispatch__
    errors = self.event_handler._handle_kernel_launch(
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 331, in _handle_kernel_launch
    stack_trace = traceback.format_stack()

Tensor was allocated with stack trace:
  File "/private/home/sypniewski/pytorch/tester.py", line 7, in <module>
    a = torch.rand(4, 2, device="cuda")
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 483, in __torch_dispatch__
    outputs = func(*args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/_ops.py", line 60, in __call__
    return self._op(*args, **kwargs or {})
  File "/private/home/sypniewski/pytorch/torch/utils/_cuda_trace.py", line 23, in fire_callbacks
    cb(*args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 383, in _handle_memory_allocation
    self.tensors_accessed.create_tensor(data_ptr, traceback.format_stack())
```




[ghstack-poisoned]

sypneiwski committed:
Update on "Add CUDA Sanitizer"

Example of a simple synchronization error:
```
a = torch.rand(4, 2, device="cuda")

with torch.cuda.stream(second_stream):
    torch.mul(a, 5, out=a)
```
Output produced by CSAN:
```
============================
CSAN detected a possible data race on tensor with data pointer 139719969079296
Access by stream 94646435460352 during kernel:
aten::mul.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
writing to argument: self, out, output
With stack trace:
  File "/private/home/sypniewski/pytorch/tester.py", line 10, in <module>
    torch.mul(a, 5, out=a)
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 486, in __torch_dispatch__
    errors = self.event_handler._handle_kernel_launch(
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 331, in _handle_kernel_launch
    stack_trace = traceback.format_stack()

Previous access by stream 0 during kernel:
aten::rand(int[] size, *, int? dtype=None, int? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
writing to argument: output
With stack trace:
  File "/private/home/sypniewski/pytorch/tester.py", line 7, in <module>
    a = torch.rand(4, 2, device="cuda")
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 486, in __torch_dispatch__
    errors = self.event_handler._handle_kernel_launch(
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 331, in _handle_kernel_launch
    stack_trace = traceback.format_stack()

Tensor was allocated with stack trace:
  File "/private/home/sypniewski/pytorch/tester.py", line 7, in <module>
    a = torch.rand(4, 2, device="cuda")
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 483, in __torch_dispatch__
    outputs = func(*args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/_ops.py", line 60, in __call__
    return self._op(*args, **kwargs or {})
  File "/private/home/sypniewski/pytorch/torch/utils/_cuda_trace.py", line 23, in fire_callbacks
    cb(*args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 383, in _handle_memory_allocation
    self.tensors_accessed.create_tensor(data_ptr, traceback.format_stack())
```




[ghstack-poisoned]

sypneiwski committed:
Update on "Add CUDA Sanitizer"

Example of a simple synchronization error:
```
a = torch.rand(4, 2, device="cuda")

with torch.cuda.stream(second_stream):
    torch.mul(a, 5, out=a)
```
Output produced by CSAN:
```
============================
CSAN detected a possible data race on tensor with data pointer 139719969079296
Access by stream 94646435460352 during kernel:
aten::mul.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
writing to argument: self, out, output
With stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 364, in _handle_kernel_launch
    stack_trace = traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 544, in __torch_dispatch__
    errors = self.event_handler._handle_kernel_launch(
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 9, in <module>
    torch.mul(a, 5, out=a)

Previous access by stream 0 during kernel:
aten::rand(int[] size, *, int? dtype=None, int? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
writing to argument: output
With stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 364, in _handle_kernel_launch
    stack_trace = traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 544, in __torch_dispatch__
    errors = self.event_handler._handle_kernel_launch(
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 6, in <module>
    a = torch.rand(10000, device="cuda")

Tensor was allocated with stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 420, in _handle_memory_allocation
    traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/utils/_cuda_trace.py", line 23, in fire_callbacks
    cb(*args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/_ops.py", line 60, in __call__
    return self._op(*args, **kwargs or {})
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 541, in __torch_dispatch__
    outputs = func(*args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 6, in <module>
    a = torch.rand(10000, device="cuda")
```




[ghstack-poisoned]

ezyang commented:
You're still going to need a user manual page, but I suppose that can come later.

sypneiwski committed:
Update on "Add CUDA Sanitizer"

Example of a simple synchronization error:
```
a = torch.rand(4, 2, device="cuda")

with torch.cuda.stream(second_stream):
    torch.mul(a, 5, out=a)
```
Output produced by CSAN:
```
============================
CSAN detected a possible data race on tensor with data pointer 139719969079296
Access by stream 94646435460352 during kernel:
aten::mul.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
writing to argument: self, out, output
With stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 364, in _handle_kernel_launch
    stack_trace = traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 544, in __torch_dispatch__
    errors = self.event_handler._handle_kernel_launch(
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 9, in <module>
    torch.mul(a, 5, out=a)

Previous access by stream 0 during kernel:
aten::rand(int[] size, *, int? dtype=None, int? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
writing to argument: output
With stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 364, in _handle_kernel_launch
    stack_trace = traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 544, in __torch_dispatch__
    errors = self.event_handler._handle_kernel_launch(
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 6, in <module>
    a = torch.rand(10000, device="cuda")

Tensor was allocated with stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 420, in _handle_memory_allocation
    traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/utils/_cuda_trace.py", line 23, in fire_callbacks
    cb(*args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/_ops.py", line 60, in __call__
    return self._op(*args, **kwargs or {})
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 541, in __torch_dispatch__
    outputs = func(*args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 6, in <module>
    a = torch.rand(10000, device="cuda")
```




[ghstack-poisoned]

sypneiwski commented:
As discussed, since the error is now raised, I've made the error class inspectable. I've also made a wrapper class for the list of errors that prints its length.

FYI @lw @bottler

sypneiwski committed:
Update on "Add CUDA Sanitizer"

Example of a simple synchronization error:
```
a = torch.rand(4, 2, device="cuda")

with torch.cuda.stream(second_stream):
    torch.mul(a, 5, out=a)
```
Output produced by CSAN:
```
============================
CSAN detected a possible data race on tensor with data pointer 139719969079296
Access by stream 94646435460352 during kernel:
aten::mul.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
writing to argument: self, out, output
With stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 364, in _handle_kernel_launch
    stack_trace = traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 544, in __torch_dispatch__
    errors = self.event_handler._handle_kernel_launch(
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 9, in <module>
    torch.mul(a, 5, out=a)

Previous access by stream 0 during kernel:
aten::rand(int[] size, *, int? dtype=None, int? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
writing to argument: output
With stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 364, in _handle_kernel_launch
    stack_trace = traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 544, in __torch_dispatch__
    errors = self.event_handler._handle_kernel_launch(
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 6, in <module>
    a = torch.rand(10000, device="cuda")

Tensor was allocated with stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 420, in _handle_memory_allocation
    traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/utils/_cuda_trace.py", line 23, in fire_callbacks
    cb(*args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/_ops.py", line 60, in __call__
    return self._op(*args, **kwargs or {})
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 541, in __torch_dispatch__
    outputs = func(*args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 6, in <module>
    a = torch.rand(10000, device="cuda")
```




[ghstack-poisoned]

sypneiwski committed:
Update on "Add CUDA Sanitizer"

Example of a simple synchronization error:
```
a = torch.rand(4, 2, device="cuda")

with torch.cuda.stream(second_stream):
    torch.mul(a, 5, out=a)
```
Output produced by CSAN:
```
============================
CSAN detected a possible data race on tensor with data pointer 139719969079296
Access by stream 94646435460352 during kernel:
aten::mul.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
writing to argument: self, out, output
With stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 364, in _handle_kernel_launch
    stack_trace = traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 544, in __torch_dispatch__
    errors = self.event_handler._handle_kernel_launch(
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 9, in <module>
    torch.mul(a, 5, out=a)

Previous access by stream 0 during kernel:
aten::rand(int[] size, *, int? dtype=None, int? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
writing to argument: output
With stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 364, in _handle_kernel_launch
    stack_trace = traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 544, in __torch_dispatch__
    errors = self.event_handler._handle_kernel_launch(
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 6, in <module>
    a = torch.rand(10000, device="cuda")

Tensor was allocated with stack trace:
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 420, in _handle_memory_allocation
    traceback.StackSummary.extract(
  File "/private/home/sypniewski/pytorch/torch/utils/_cuda_trace.py", line 23, in fire_callbacks
    cb(*args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/_ops.py", line 60, in __call__
    return self._op(*args, **kwargs or {})
  File "/private/home/sypniewski/pytorch/torch/cuda/_sanitizer.py", line 541, in __torch_dispatch__
    outputs = func(*args, **kwargs)
  File "/private/home/sypniewski/pytorch/torch/utils/_python_dispatch.py", line 76, in wrapped
    return f(self, *args, **kwargs)
  File "/private/home/sypniewski/pytorch/tester.py", line 6, in <module>
    a = torch.rand(10000, device="cuda")
```




[ghstack-poisoned]

sypneiwski commented:
@pytorchbot merge -g
